lib/streamlit/watcher/event_based_path_watcher.py

```diff
@@ -359,19 +359,24 @@ def handle_path_change_event(self, event: events.FileSystemEvent) -> None:
 
         abs_changed_path = os.path.abspath(changed_path)
 
-        # First check if the exact path is being watched
-        changed_path_info = self._watched_paths.get(abs_changed_path, None)
-
-        # If the exact path isn't found, check if it's inside any watched directories
-        # This is necessary for the folder watching feature to detect changes to files
-        # within watched directories, not just the directories themselves
-        for path, info in self._watched_paths.items():
-            if (
-                os.path.isdir(path)
-                and os.path.commonpath([path, abs_changed_path]) == path
-            ):
-                changed_path_info = info
-                break
+        # To prevent a race condition, we hold a lock while accessing
+        # _watched_paths.
+        with self._lock:
+            # First check if the exact path is being watched
+            changed_path_info = self._watched_paths.get(abs_changed_path, None)
+
+            # If the exact path isn't found, check if it's inside any watched
+            # directories. This is necessary for the folder watching feature to
+            # detect changes to files within watched directories, not just the
+            # directories themselves.
+            if changed_path_info is None:
```

Comment:
Another related error, previously we always run the loop... but we should only run it when `changed_path_info is None`